io/src/main/scala/fs2/io/udp/AsynchronousSocketGroup.scala

```diff
@@ -0,0 +1,146 @@
+    readers: MutableQueue[Either[Throwable,Packet] => Unit] = MutableQueue(),
+    writers: MutableQueue[(Packet,Option[Throwable] => Unit)] = MutableQueue()
+  ) {
+
+    def dequeueReader: Option[Either[Throwable,Packet] => Unit] = {
+      if (readers.isEmpty) None
+      else Some(readers.dequeue())
+    }
+
+    def queueReader(reader: Either[Throwable,Packet] => Unit): Unit = {
+      readers += reader
+      ()
+    }
+
+    def hasReaders: Boolean = readers.nonEmpty
+
+    def peekWriter: Option[(Packet,Option[Throwable] => Unit)] = {
+      if (writers.isEmpty) None
+      else Some(writers.head)
+    }
+
+    def dequeueWriter: Option[(Packet,Option[Throwable] => Unit)] = {
+      if (writers.isEmpty) None
+      else Some(writers.dequeue())
+    }
+
+    def queueWriter(writer: (Packet,Option[Throwable] => Unit)): Unit = {
+      writers += writer
+      ()
+    }
+
+    def hasWriters: Boolean = writers.nonEmpty
+
+    def close: Unit = {
+      readers.foreach { cb => cb(Left(new ClosedChannelException)) }
+      readers.clear
+      writers.foreach { case (_, cb) => cb(Some(new ClosedChannelException)) }
+      writers.clear
+    }
+  }
+
+  def apply(): AsynchronousSocketGroup = new AsynchronousSocketGroup {
+
+    private val selectorLock = new ReentrantLock()
+    private val selector = Selector.open()
+    private val pendingThunks: MutableQueue[() => Unit] = MutableQueue()
+
+    override def register(channel: DatagramChannel): (AsynchronousSocketGroup.Attachment, SelectionKey) = {
+      channel.configureBlocking(false)
+      val attachment = new AsynchronousSocketGroup.Attachment()
+      val key = {
+        selectorLock.lock
+        try {
+          selector.wakeup
+          channel.register(selector, 0, attachment)
+        } finally {
+          selectorLock.unlock
+        }
+      }
+      (attachment, key)
+    }
+
+    override def enqueue(f: => Unit): Unit = {
+      pendingThunks.synchronized {
+        pendingThunks.+=(() => f)
+      }
+      selector.wakeup()
+      ()
+    }
+
+    private val doneNow = new AtomicBoolean(false)
+    private val selectorThread = new Thread(new Runnable {
+      def run = try {
+        val readBuffer = ByteBuffer.allocate(1 << 16)
+        while (!doneNow.get && !Thread.currentThread.isInterrupted) {
+          pendingThunks.synchronized { while (pendingThunks.nonEmpty) pendingThunks.dequeue()() }
+          selectorLock.lock
+          selectorLock.unlock
+          println(selector.select)
+          val selectedKeys = selector.selectedKeys.iterator
+          while (selectedKeys.hasNext) {
+            val key = selectedKeys.next
+            selectedKeys.remove
+            val channel = key.channel.asInstanceOf[DatagramChannel]
+            val attachment = key.attachment.asInstanceOf[Attachment]
+            if (key.isValid) {
+              if (key.isReadable) {
+                readBuffer.clear
+                val src = channel.receive(readBuffer)
```

Comment:
agreed like the read shall be almost non-blocking. As for the send however I think we have to handle IOException there